pgjdbc/src/main/java/org/postgresql/util/HostSpec.java

```diff
@@ -39,4 +46,50 @@ public boolean equals(Object obj) {
   public int hashCode() {
     return port ^ host.hashCode();
   }
+
+  public Boolean shouldResolve() {
+    if (System.getProperty("socksProxyHost") == null) {
+      return true;
+    }
+    return matchesNonProxyHosts();
+  }
+
+  private Boolean matchesNonProxyHosts() {
+    String nonProxyHosts = System.getProperty("socksNonProxyHosts", DEFAULT_NON_PROXY_HOSTS);
+    if (nonProxyHosts == null || this.host.isEmpty()) {
+      return false;
+    }
+
+    Pattern pattern = toPattern(nonProxyHosts);
+    Matcher matcher = pattern == null ? null : pattern.matcher(this.host);
+    return matcher != null && matcher.matches();
+  }
+
+  private Pattern toPattern(String mask) {
+    StringBuilder joiner = new StringBuilder();
+    String separator = "";
+    for (String disjunct : mask.split("\\|")) {
+      if (!disjunct.isEmpty()) {
+        String regex = disjunctToRegex(disjunct.toLowerCase());
+        joiner.append(separator).append(regex);
+        separator = "|";
+      }
+    }
+
+    return joiner.length() == 0 ? null : compile(joiner.toString());
+  }
+
+  private String disjunctToRegex(String disjunct) {
+    String regex;
+
+    if (disjunct.startsWith("*")) {
+      regex = ".*" + Pattern.quote(disjunct.substring(1));
+    } else if (disjunct.endsWith("*")) {
```

Comment:
Should `*hostname*` kind of patterns be supported?

Reply:
I'm replicating exactly what the JVM does in `sun.net.spi.DefaultProxySelector`, and it does not support that pattern. So I do not think we should (because we would resolve a host and then the JVM would send it through the proxy anyways).